proof/refine/ARM/Refine.thy

```diff
@@ -229,19 +229,29 @@ lemma absKState_correct:
 text \<open>The top-level invariance\<close>
 
 lemma activate_thread_sched_act:
-  "\<lbrace>ct_in_state activatable and
-    (\<lambda>s. is_schedulable_bool (cur_thread s) s \<and> P (scheduler_action s))\<rbrace>
+  "\<lbrace>ct_in_state activatable and (\<lambda>s. is_schedulable_bool (cur_thread s) s) and schact_is_rct\<rbrace>
    activate_thread
-   \<lbrace>\<lambda>_ s. P (scheduler_action s)\<rbrace>"
+   \<lbrace>\<lambda>_. schact_is_rct\<rbrace>"
   unfolding activate_thread_def
-  sorry (*
-  by (simp add: activate_thread_def set_thread_state_def arch_activate_idle_thread_def
-      | (wp set_thread_state_sched_act gts_wp)+ | wpc)+ *)
+  apply (wpsimp wp: set_thread_state_schact_is_rct_strong gts_wp hoare_drop_imp hoare_vcg_all_lift)
+  by (clarsimp simp: is_schedulable_bool_def2)
+
+crunches sc_and_timer
+  for scheduler_action[wp]: "\<lambda>s. P (scheduler_action s)"
+  (wp: crunch_wps)
+
+lemmas ssa_schact_is_rct_obvious[wp] = set_scheduler_action_obvious [
```

Comment:
Nitpick: did you mean to have the square bracket at the end of the line?

Reply:
I did, but I can change it if that looks wrong.